zerver/lib/upload.py

```diff
@@ -149,6 +149,11 @@ def get_file_info(request, user_file):
     # type: (HttpRequest, File) -> Tuple[text_type, Optional[text_type]]
 
     uploaded_file_name = user_file.name
+    if not isinstance(user_file.name, str):
+        # This check makes sure that the type of the file name is not
+        # Python 2 Unicode.
+        uploaded_file_name = uploaded_file_name.encode('ascii')
```

Comment:
There are definitely tradeoffs here.  I agree with the reasoning behind making the library more flexible, in the sense that it could protect us from future bugs, and it's only like two lines of code.  But I stand behind my original assertion that we should validate/coerce types at the boundary of the system.  If the outside world gives us a filename in some type or encoding that we don't like, we should convert it or reject it at the outer boundaries of our system.

@umairwaheed To your concern about future bugs, which I think is valid, I think we can address it as follows:

- lock down the mypy annotations
- add a runtime assert on the incoming type
- rely on the fact that any future endpoint will have an end-to-end test (due to code review stringency which you just witnessed, plus eventual lint/test-based lockdown on untested endpoints, which is pretty imminent)
- python3 is more generally sane about unicode